### DISCUSSION TITLE
Leader and managers recovery

### DISCUSSION THREAD

**Discussion by trafficavenue** [caprover/caprover#1029] [Q&A]
Hi everyone!

There is something I don't understand : if everything lives in the leader node (caprover, nginx etc...) how are we supposed to keep the cluster online if the leader node fails and we have managers and workers? Swarm will elect a new leader but since everything is setup inside the leader node I don't understand how to keep the cluster alive.

Thanks!

Alex

**Comment by githubsaturn** [comment]
You are correct. Leader node is the the single point of failure. 
https://github.com/caprover/caprover/issues/162
https://github.com/caprover/caprover/issues/197

**Reply by trafficavenue** [reply]
Got it. So this also mean that if I have VPS located in NY and San Francisco under the same cluster and the leader node (where captain lives) is located in Amsterdam, all the websites hosted on the cluster will be actually located in Amsterdam captain acting like a proxy to access the content of each node? I place each node very far away for the sake of the discussion but that also means the inter-cluster connectivity is very important for the cluster to be efficient.

**Reply by githubsaturn** [reply]
Your website app might be in leader node, might be on worker node. If it's on worker node, a request will follow this path:
```
user ===>> leader node ====>> worker node
```

**Reply by trafficavenue** [reply]
Got it, thanks! 

One more question (sorry!) : did you tried mounting a network disk on each member of the cluster and uses the Docker mountpoint to connect to it? That would indeed make the cluster persistent since the data will be available for each node. If I want to do the same with captain (and have captain available cluster wide to remove the SPOF) what should I change?

**Reply by githubsaturn** [reply]
You also have your DNS pointing to the leader node. A completely bulletproof resilient system is not something that you can achieve automatically. If your main node goes down, even if you have mapped `/captain` directory to another node, your DNS still points to the main node, so you'd have to
- Update your DNS to point to the other node.
- Change the node id constraint on docker service to move the `captain-captain` service to the new node.

**Reply by trafficavenue** [reply]
Thanks,

Not really related but since CapRover doesn't allow for arguments while mounting a disk mount I have created a NFS mount directly on the host (ubuntu side) and using it inside CapRover like a local bind mount for the containers. Is is something that you think will be working at scale?

**Reply by githubsaturn** [reply]
I am not sure, I haven't tried it. But feel free to post your results here once you find out. I am curious to know if it works.

**Reply by trafficavenue** [reply]
So here's what I tried (so far) : 

1/ I've launched an independent VM with 100GB SSD disk in the same datacenter and installed NFS and created /var/capdata
2/ I've mounted the /var/capdata into two member of the swarm (1 leader node with caprover, 1 worker) and created one subfolder by app. I've also installed cachefilesd to cache the files regularly used and avoid over usage of the NFS server.
3/ Inside the app settings I've selected a bind mount to /var/capdata/appXX and deployed Wordpress, MySQL and some other apps to different containers

The results are pretty good (and I can move in one click an app from a node to another without dataloss) but I still have to try to stress the system to see how it affects performance.

That being said from my research, and with like any network based file system, it's important to have in production : 

SSD disks both in the NFS server and clients
A LOT of RAM (at least 128GB) on the NFS server
A good connectivity between server and client (ideally a 10Gbps internal network)

If you have specific tests you want to run or ideas to stress the system don't hesitate!

**Reply by githubsaturn** [reply]
This is cool! Feel free to write up a tutorial/guide/blogpost and we'll link it in the official doc. 

But yea, generally, NFS comes with a lot of overhead. Generally speaking, recurrent backups are usually the way to go. But it's nice to know that NFS works fine as well if someone wants it
